order_book/src/order.rs

```diff
@@ -0,0 +1,303 @@
+                // is in the same token id as the base token, since it takes
+                // away from the amount of base tokens available for consumption
+                // by the fulfiller.
+                if num_required_outputs == 1 {
+                    if sci.required_output_amounts[0].token_id != base_token_id {
+                        return Err(Error::UnsupportedSci(format!(
+                        "Suspected required-change-output token id {} does not match partial output token id {}",
+                        sci.required_output_amounts[0].token_id, amount.token_id
+                    )));
+                    }
+                    max_base_amount = max_base_amount
+                        .checked_sub(sci.required_output_amounts[0].value)
+                        .ok_or_else(|| {
+                            Error::UnsupportedSci(format!(
+                                "max base amount {} is lower than required change {}",
+                                max_base_amount, sci.required_output_amounts[0].value
+                            ))
+                        })?;
+                }
+
+                (
+                    amount.token_id,
+                    min_base_amount..=max_base_amount,
+                    amount.value,
+                )
+            }
+            _ => {
+                return Err(Error::UnsupportedSci(format!(
+                    "Unsupported number of required/partial outputs {}/{}",
+                    input_rules.required_outputs.len(),
+                    input_rules.partial_fill_outputs.len()
+                )))
+            }
+        };
+
+        let id = OrderId::from(&sci);
+
+        let pair = Pair {
+            base_token_id,
+            counter_token_id,
+        };
+
+        Ok(Self {
+            sci,
+            id,
+            pair,
+            base_range,
+            max_counter_tokens,
+            timestamp,
+        })
+    }
+
+    /// Get underlying SCI.
+    pub fn sci(&self) -> &SignedContingentInput {
+        &self.sci
+    }
+
+    /// Get unique identifier.
+    pub fn id(&self) -> &OrderId {
+        &self.id
+    }
+
+    /// Get the pair being traded by this order.
+    pub fn pair(&self) -> &Pair {
+        &self.pair
+    }
+
+    /// Get the range of base tokens offered by this order (the minimum and
+    /// maximum amount of base token that can be obtained by fulfiling the
+    /// order).
+    pub fn base_range(&self) -> &RangeInclusive<u64> {
+        &self.base_range
+    }
+
+    /// Get the maximum amount of base tokens that can be provided by this
+    /// order.
+    pub fn max_base_tokens(&self) -> u64 {
+        *self.base_range.end()
+    }
+
+    /// Get the maximum amount of counter tokens required to completely use all
+    /// the available base tokens
+    pub fn max_counter_tokens(&self) -> u64 {
+        self.max_counter_tokens
+    }
+
+    // Get the number of counter tokens we will need to provide in order to consume
```

Comment:
Would it provide more clarity to specify that it's the amount of counter tokens provided by the counterparty?